state/state.go

```diff
@@ -119,7 +121,24 @@ func NewWithStore(token string, cabinet store.Cabinet) (*State, error) {
 		return nil, err
 	}
 
-	return NewFromSession(s, cabinet), nil
+	return newWithAutoRescale(s, cabinet), nil
+}
+
+func newWithAutoRescale(s *session.Session, cabinet store.Cabinet) *State {
+	state := NewFromSession(s, cabinet)
+	state.ShardManager.Rescale = func() *shard.Manager {
+		state.Reset()
```

Comment:
`Reset` is not only called here but also on `Ready` events, which would cause the state to be incomplete when the shards are starting up. A way to fix this would probably be to have a `noResetOnReady bool` that prevents the handlers from resetting the state there, so we can do it here.

This call should also be moved to after the `NewManager` call.

Reply:
If we expect people to implement their own shard manager, we should probably expose `noResetOnReady`.